added page, which displays albums added in user's library;

OverKrice3000 opened the pull request:
added functions to query some albums, display them, add or remove to user's library;
minor styles addings / improvements;
changed font to courier;
index and library pages are functional;

OverKrice3000 committed:
added page, which displays albums added in user's library;
added functions to query some albums, display them, add or remove to user's library;
minor styles addings / improvements

alserz1 commented on package.json:
```diff
@@ -10,6 +10,7 @@
     "@types/node": "^16.11.22",
     "@types/react": "^17.0.39",
     "@types/react-dom": "^17.0.11",
+    "axios": "^0.27.2",
```
axios вроде нигде в проекте не используется, зачем он здесь?

OverKrice3000 replied:
сначала думал его использовать, потом подумал что надо наверное fetch api. Удалил его

alserz1 commented on public/js/album/album_api.js:
```diff
@@ -0,0 +1,98 @@
+ * This file contains functions, that query external spotify api and return Promise
+ */
+
+import * as auth from "../auth/auth.js";
```
import * лучше не делать. Некоторые инструменты могут анализировать код и выкидывать неиспользуемые части, а import * мешает это делать

OverKrice3000 replied:
поменял на импорт конкретных функций

alserz1 commented on public/js/album/album_api.js:
```diff
@@ -0,0 +1,98 @@
+ * This function queries new releases
+ * @param offset an offset, from which albums will be queried
+ * @param limit maximum amount of albums, which will be received in response
+ * @returns {Promise<any>} a promise with list of new releases
```
any никогда не должно использоваться в типах, особенно там, где ты знаешь реальный тип

OverKrice3000 replied:
Поменял на реальный тип

alserz1 commented on public/js/album/album_api.js:
```diff
@@ -0,0 +1,98 @@
+    return fetch("https://api.spotify.com/v1/browse/new-releases", {
+        method: "GET",
+        headers: {
+            "Authorization": `Bearer ${auth.access_token}`,
```
как я понимаю, токен закэшен и обновляется только в очень маленьком количестве ситуаций. Тут каждый метод может упасть с ошибкой из-за того, что срок действия токена истёк и это никак не обрабатывается. Да и вообще обработки ошибок нет

OverKrice3000 replied:
Сделал обработку ошибок. Раньше access token менялся при загрузке страниц, теперь он меняется еще когда возникает ошибка. Он по большому счету и не хардкодился, в отличии от refresh token, в переменной просто лежало какое то старое значение. 
Мы же не можем поставить куку с access token если не запускаем сервер? LocalStorage вроде с file: дает неопределенное поведение. Вот и остается хранить в переменной
refresh token я так понял получать автоматически не выйдет, потому что спотифай сначала просит пользователя подтвердить, а потом отправляет callback, который некому обрабатывать, т.к. сервера нет. Поэтому остается только хардкодить.

alserz1 replied:
> Мы же не можем поставить куку с access token если не запускаем сервер?

если честно не помню, проще взять да проверить) Там в ответе сервера ещё есть поле `expires_in`, можно просто таймер ставить. Но таймеры это вещь очень непредсказуемая, т.к. браузер может их замедлять или вообще отключать, поэтому есть другой вариант - spotify наверное будет ругаться, что токен истёк, так что если поймали такую ошибку – запрашиваем новый токен и повторяем запрос.

Про `callback` и `refresh_token` не очень понял. Там `redirect_uri` требуется в любом случае при твоей схеме

alserz1 commented on public/js/album/album_component_builders.js:
```diff
@@ -0,0 +1,74 @@
+
+    let music_set_name = document.createElement("h2")
+    music_set_name.classList.add("music-set__name")
+    music_set_name.innerHTML = `${album.name}`
```
вставлять динамический контент через innerHTML это плохая идея: https://developer.mozilla.org/en-US/docs/Web/API/Element/innerHTML#security_considerations

OverKrice3000 replied:
поменял на textContent

alserz1 commented on public/js/album/album_funcs.js:
```diff
@@ -0,0 +1,89 @@
+ */
+export function preprocess_saved_albums_received(saved_albums){
+    let albums = []
+    for(let i = 0; i < saved_albums.length; i++){
```
вся функция это `return saved_albums.map((album) => album.album);`

я бы это даже в отдельную функцию не выделял, это используется в одном месте, кода очень мало и никакой особой смысловой нагрузки он не несёт. Только лишняя беготня по файлу

и вкусовщина, но я вообще не люблю метод for, потому что его каждый раз нужно полностью читать, чтобы понять замысел автора. Всякие map/filter/find/forEach и т.п. куда проще воспринимать

OverKrice3000 replied:
заинлайнил функцию

alserz1 commented on public/js/album/album_api.js:
```diff
@@ -0,0 +1,98 @@
+ * @param limit maximum amount of albums, which will be received in response
+ * @returns {Promise<any>} a promise with list of new releases
+ */
+export async function get_new_releases(offset, limit){
```
в js принят camelCase для именования функций, но настаивать на переименовании я не буду

OverKrice3000 replied:
поменяю тогда наверное на реакте, там все равно код рефакторить надо будет)

alserz1 commented on public/js/auth/auth.js:
```diff
@@ -0,0 +1,28 @@
+
+let refresh_token = 'AQC6TQh_U7lMCJgWZAudN6LqdGYU7v01e7fhmrT_Qf2hZQ8W6Jb7QgcN6nPxBwWfNKSLW9kFJ4w17dxRm0ssD7njgQ7eKN3MYei2CqK1F1Ldsec-SrPHRTBLUSzmxHnv6XQ'
```
оба этих токена могут устареть, нельзя их хардкодить

OverKrice3000 replied:
^
|||
|||

alserz1 commented on public/index.html:
```diff
@@ -11,6 +11,7 @@
       name="description"
       content="Web site created using create-react-app"
     />
+    <script src="js/index.js" type="module" defer></script>
```
defer не имеет смысла на скриптах с type="module", они и так defer

OverKrice3000 replied:
убрал

OverKrice3000 committed:
removed axios from dependencies;
changed import statements, removing import *;
Changed any in Promise return types;
removed defer from script tag descriptions;
inlined preprocess_saved_albums function;
changed innerHTML to textContent;

OverKrice3000 committed:
removed axios from dependencies;
changed import statements, removing import *;
Changed any in Promise return types;
removed defer from script tag descriptions;
inlined preprocess_saved_albums function;
changed innerHTML to textContent;

OverKrice3000 committed:
added error handling when access token is invalid or expired

OverKrice3000 committed:
added access token update on library page loading

alserz1 commented on public/js/album/album_api.js:
```diff
@@ -0,0 +1,142 @@
+ * This function queries new releases
+ * @param offset an offset, from which albums will be queried
+ * @param limit maximum amount of albums, which will be received in response
+ * @returns {Promise<Object>} a promise with list of new releases
```
object не сильно полезнее, чем any, потому что в js вообще всё является объектами, если глубоко копать) Но на описаниях типов я настаивать не буду, иначе никто диплом не сдаст)

в большинстве мест тип был бы описан через typescript. Но можно и @typedef сделать через JSDoc, описать там форму объекта

alserz1 commented on public/js/album/album_api.js:
```diff
@@ -0,0 +1,142 @@
+ * @returns {Promise<Object>} a promise with list of new releases
+ */
+export async function get_new_releases(offset, limit){
+    let new_releases_fetch_query = () => fetch("https://api.spotify.com/v1/browse/new-releases", {
```
там сейчас пойдёт копипаста с fetch и обработкой ошибок, это же можно вынести в отдельное место
